### ISSUE TITLE
Proposal: Modules

### ISSUE THREAD

**Issue by timsuchanek** [graphcool/graphcool-cli#162]
## Problem
We want project definitions to be reusable.
Especially when you want to pull different "packages" or how we currently call them "examples", where specific problems like auth have been solved, we need a concept to tie them together into one project, while still having the flexibility of being able to change the code.

### Modules
Introducing the `module` concept.
A module is a folder that contains a `graphcool.yml` definition and serveral type definitions.
A project then contains a `modules` folder that contains multiple sub-modules it's using. A project can be seen as a module, too.
In the beginning, only a flat hierachy will be allowed, later maybe also a deeper hierachy.
Example project structure:
```
.
├── .graphcoolrc
├── code
│   └── filter-posts.js
├── graphcool.yml
├── modules
│   ├── logging
│   │   ├── code
│   │   │   ├── log-posts.graphql
│   │   │   └── log-posts.js
│   │   ├── graphcool.yml
│   │   └── types.graphql
│   └── weather
│       ├── code
│       │   ├── weather.graphql
│       │   └── weather.js
│       ├── graphcool.yml
│       └── types.graphql
└── types.graphql
```
### graphcool.yml definition
```yaml
modules:
 my: this/is/my/module/path
 sane: modules/sane-path
# stuff: modules/not-using-now
```
### Distribution
Modules could be installed via the following command:
```sh
$ gc module add graphcool/weather-example
```
Where `graphcool/weather-example` is refering to the repo on github
Maybe later we want to support a command like this:
```sh
$ gc module add graphcool/examples/some/deep/folder
```
With that you could install a module from a specific folder from a github repo.

### Type Merging
- Will it be possible to merge types? For example adding a `token: String!` field to the `User` type for auth
### Versioning
We need to introduce the concept of a version of a module. We could for example reuse the `package.json`, which is already a well-thought format for package definitions. The question is, if we have to reinvent the wheel here, or could reuse existing package formats. However, the version is essential in order for developers to update modules gracefully.

### Backend
- Merge modules into one running project

### CLI
- pull in modules from `modules` folder
- implement `module add`, `module remove`

**Comment by kbrandwijk** [comment]
In addition, it would be great if a module could be installed as npm module, that exports the required elements. The module could then be defined like dependencies in package.json, but in the graphcool.yml instead.

**Comment by timsuchanek** [comment]
Closing it for now, as a basic version is implemented in 1.3.13
